Composite/Composite/MComp.fs

```diff
@@ -0,0 +1,47 @@
+namespace Composite
+
+open FSharp.Core
+
+[<RequireQualifiedAccess>]
+module MComp =
+
+    ///<summary>Initializes an empty marked composite.</summary>
+    ///<param name="mark">The mark.</param>
+    ///<typeparam name="TMark">The type of marks in the composite.</typeparam>
+    ///<typeparam name="TPayload">The type of payload objects in the composite.</typeparam>
+    let empty<'TMark, 'TPayload> (mark: 'TMark) =
+        MarkedComposite {Mark = mark; Components = Seq.empty<Composite<'TMark, 'TPayload>>}
+
+    ///<summary>Returns the sequence of components of the input marked composite.</summary>
+    ///<param name="source">The input composite.</param>
+    ///<typeparam name="TMark">The type of marks in the composite.</typeparam>
+    ///<typeparam name="TPayload">The type of payload objects in the composite.</typeparam>
+    let components (source: Composite<'TMark, 'TPayload>) =
+        match source with
+        | MarkedComposite x -> x.Components
+        | x -> seq { yield x }
+
+    ///<summary>Builds a new marked composite based on the source in which the value is set at the given mark in the given container.</summary>
+    ///<param name="containerMark">The mark of the container in which the value should be set.</param>
+    ///<param name="valueMark">The mark to set the value at.</param>
+    ///<param name="value">The value to set.</param>
+    ///<param name="markComparer">The function used to determine if two marks match.</param>
+    ///<param name="source">The input composite.</param>
+    ///<typeparam name="TMark">The type of marks in the composite.</typeparam>
+    ///<typeparam name="TPayload">The type of payload objects in the composite.</typeparam>
+    let rec setValue (containerMark: 'TMark) (valueMark: 'TMark) (value: 'TPayload) (markComparer: 'TMark -> 'TMark -> bool) (source: Composite<'TMark, 'TPayload>) =
+        match source with
+        | MarkedComposite x ->  if x.Mark |> markComparer containerMark then
+                                    MarkedComposite {Mark = x.Mark; Components = seq{
+                                        let mutable valueFound = false
```

Comment:
At first look I thought that we do not need to call mark comparer if value already has been found. Probably we can get rid of mutable value by creating a recursive function with some kind of an accumulator which will has the same `valueFound` parameter and if it is true then do yield! a components tail.

Reply:
Now, I kinda hesitate to use recursion here. This would likely complicate the algorithm. See the [yield! study case](https://github.com/vkamiansky/composite/wiki/Problem-2:-yield!).